spring-batch-core/src/test/java/org/springframework/batch/core/launch/support/TaskExecutorJobOperatorTests.java

```diff
@@ -427,6 +428,52 @@ void testAbortNonStopping() {
 		assertThrows(JobExecutionAlreadyRunningException.class, () -> jobOperator.abandon(123L));
 	}
 
+	@Test
+	void testRecover() {
+		JobInstance jobInstance = new JobInstance(123L, job.getName());
+		JobExecution jobExecution = new JobExecution(jobInstance, 111L, jobParameters);
+		jobExecution.setStatus(BatchStatus.STARTED);
+		jobExecution.createStepExecution("step1").setStatus(BatchStatus.STARTED);
+		when(jobRepository.getJobExecution(111L)).thenReturn(jobExecution);
+		when(jobRepository.getLastJobExecution(jobInstance)).thenReturn(jobExecution);
+		JobExecution recover = jobOperator.recover(jobExecution);
+		assertEquals(BatchStatus.FAILED, recover.getStatus());
+		assertTrue(recover.getExecutionContext().containsKey("recovered"));
```

Comment:
We need to asset that the end time is not null as well. I will take care of adding it.